crates/ide_assists/src/handlers/destructure_tuple_binding.rs

```diff
@@ -0,0 +1,2141 @@
+) -> String {
+    // FIXME: detect if name already used
+    format!("_{}", index)
+}
+
+enum RefType {
+    ReadOnly,
+    Mutable,
+}
+struct TupleData {
+    ident_pat: IdentPat,
+    // name: String,
+    range: TextRange,
+    ref_type: Option<RefType>,
+    field_names: Vec<String>,
+    // field_types: Vec<Type>,
+    usages: Option<UsageSearchResult>,
+}
+fn edit_tuple_assignment(
+    ctx: &AssistContext,
+    builder: &mut AssistBuilder,
+    data: &TupleData,
+    in_sub_pattern: bool,
+) {
+    let tuple_pat = {
+        let original = &data.ident_pat;
+        let is_ref = original.ref_token().is_some();
+        let is_mut = original.mut_token().is_some();
+        let fields = data.field_names.iter().map(|name| {
+            ast::Pat::from(ast::make::ident_pat(is_ref, is_mut, ast::make::name(name)))
+        });
+        ast::make::tuple_pat(fields)
+    };
+
+    let add_cursor = |text: &str| {
+        // place cursor on first tuple item
+        let first_tuple = &data.field_names[0];
+        text.replacen(first_tuple, &format!("$0{}", first_tuple), 1)
+    };
+
+    // with sub_pattern: keep original tuple and add subpattern: `tup @ (_0, _1)`
+    if in_sub_pattern {
+        let text = format!(" @ {}", tuple_pat.to_string());
+        match ctx.config.snippet_cap {
+            Some(cap) => {
+                let snip = add_cursor(&text);
+                builder.insert_snippet(cap, data.range.end(), snip);
+            }
+            None => builder.insert(data.range.end(), text),
+        };
+    } else {
+        let text = tuple_pat.to_string();
+        match ctx.config.snippet_cap {
+            Some(cap) => {
+                let snip = add_cursor(&text);
+                builder.replace_snippet(cap, data.range, snip);
+            }
+            None => builder.replace(data.range, text),
+        };
+    }
+}
+
+fn edit_tuple_usages(
+    data: &TupleData,
+    builder: &mut AssistBuilder,
+    ctx: &AssistContext,
+    in_sub_pattern: bool,
+) {
+    if let Some(usages) = data.usages.as_ref() {
+        for (file_id, refs) in usages.iter() {
+            builder.edit_file(*file_id);
+
+            for r in refs {
+                edit_tuple_usage(ctx, builder, r, data, in_sub_pattern);
+            }
+        }
+    }
+}
+fn edit_tuple_usage(
+    ctx: &AssistContext,
+    builder: &mut AssistBuilder,
+    usage: &FileReference,
+    data: &TupleData,
+    in_sub_pattern: bool,
+) {
+    match detect_tuple_index(usage, data) {
+        Some(index) => edit_tuple_field_usage(ctx, builder, data, index),
+        None => {
+            if in_sub_pattern {
+                cov_mark::hit!(destructure_tuple_call_with_subpattern);
+                return;
+            }
+
+            // no index access -> make invalid -> requires handling by user
+            // -> put usage in block comment
+            builder.insert(usage.range.start(), "/*");
+            builder.insert(usage.range.end(), "*/");
```

Comment:
Might be good to add a note about how this has side effects for macros, but us being fine with that as its unlikely to keep the macro invocation working.